include/nlohmann/json.hpp

```diff
@@ -1642,6 +1642,9 @@ class basic_json
             std::for_each(init.begin(), init.end(), [this](const detail::json_ref<basic_json>& element_ref)
             {
                 auto element = element_ref.moved_or_copied();
+#if JSON_DIAGNOSTICS
+                (*element.m_value.array)[1].m_parent = this;
```

Comment:
It's not obvious to me why this is `[1]` here.  Perhaps it deserves a comment, or perhaps like other places this could be `m_value.object->back().m_parent = this` after the following line?

Unrelated question, would this be less verbose using range-for instead of `std::for_each`?

Reply:
Would it make sense to have a `set_parent()` function that is a no-op if `JSON_DIAGNOSTICS` is not defined, and save some `#if/#endif`?

We would still want it for the loop below, but it would save one-off functions.  On the other hand, I suppose that there could also be a function that takes an iterable and sets the parent pointer of every element.

    #if JSON_DIAGNOSTICS
    set_parent_pointer(auto const &iterable) {
        for (auto& element : iterable) {
            element.m_parent = this;
        }
    }
    #else
    set_parent_pointer(auto const &) {
    }
    #endif

Reply:
I guess you'd need separate functions for array vs object.

Reply:
@gregmarr You're right - much better without `std::for_each`.

Reply:
I added a `set_parent` function. Good cleanup!